### PR TITLE
EOS-28040: [RGW] REST API to Create IAM user in RGW 

### PR DESCRIPTION
**Pull request by rohitkolapkar**
# Problem Statement
- [EOS-28040:REST API to Create IAM User](https://jts.seagate.com/browse/EOS-28040)

# Design
-  [Create-user REST API Specification](https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/884703300/Create+users+in+ceph+REST+API+Specifications)

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [x] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

# Testing Screenshots

**1. Create IAM user in RGW using csm rest api**
steps- 
a. Login using csm admin/manage role user
b. Use the bearer token to hit create iam users api - /api/v2/s3/iam/users
image
response body - 
`{"tenant": "", "user_id": "demo-user01", "display_name": "demo-user01", "email": "demouser01@seagate.com", "suspended": 0, "max_buckets": 1000, "subusers": [], "keys": [{"user": "demo-user01", "access_key": "RJL533F82YJ8TUWZUGTA", "secret_key": "dXKgC8UrmGsQXmqAxqGFy404sNtzDRbCwxOo7vAg"}], "swift_keys": [], "caps": [], "op_mask": "read, write, delete", "system": "false", "admin": "false", "default_placement": "", "default_storage_class": "", "placement_tags": [], "bucket_quota": {"enabled": false, "check_on_raw": false, "max_size": -1, "max_size_kb": 0, "max_objects": -1}, "user_quota": {"enabled": false, "check_on_raw": false, "max_size": -1, "max_size_kb": 0, "max_objects": -1}, "temp_url_keys": [], "type": "rgw", "mfa_ids": []}`

**2. Try to create already existing user, expecting error.**
image

### CODE REVIEW

**Comment by rohitkolapkar** [comment]
https://community.codacy.com/t/pep257-tool-d203-conflicts-d211/216

File: csm/common/permission_names.py [hunk 1]
```diff
@@ -36,6 +36,7 @@ class Resource:
     LYVE_PILOT = 'lyve_pilot'
     HEALTH = 'health'
     CLUSTER_MANAGEMENT = 'cluster_management'
+    RGW_S3_USERS = 'rgw_s3_users'
```
Comment (pawankumarsrivastava) [inline csm/common/permission_names.py:39]: Rename rgw_s3_users --> s3_iam_users

File: csm/core/agent/csm_agent.py [hunk 2]
```diff
@@ -165,6 +165,9 @@ def init():
         CsmRestApi._app[const.S3_ACCESS_KEYS_SERVICE] = S3AccessKeysService(s3)
         CsmRestApi._app[const.S3_SERVER_INFO_SERVICE] = S3ServerInfoService()
 
+        # RGW
```
Comment (pawankumarsrivastava) [inline csm/core/agent/csm_agent.py:168]: Write proper comments, otherwise remove it

File: csm/core/agent/csm_agent.py [hunk 3]
```diff
@@ -209,6 +212,17 @@ def _configure_cluster_management_service(message_bus_obj):
         cluster_management_service = ClusterManagementAppService(cluster_management_plugin_obj, message_bus_obj)
         CsmRestApi._app[const.CLUSTER_MANAGEMENT_SERVICE] = cluster_management_service
 
+    @staticmethod
+    def _configure_rgw_s3_users_service():
+        # ToDo: Replace the keys with consts
+        auth_user =  Conf.get(const.CSM_GLOBAL_INDEX, 'RGW>s3>iam>admin_user', None)
+        auth_user_access_key = Conf.get(const.CSM_GLOBAL_INDEX, 'RGW>s3>iam>admin_access_key', None)
+        auth_user_secrete_key = Conf.get(const.CSM_GLOBAL_INDEX, 'RGW>s3>iam>admin_secret_key', None)
+        rgw_plugin = import_plugin_module(const.RGW_PLUGIN)
```
Comment (pawankumarsrivastava) [inline csm/core/agent/csm_agent.py:221]: Rename rgw_plugin --> s3_iam_plugin
rgw_plugin_obj --> s3_iam_plugin_obj

File: csm/core/agent/csm_agent.py [hunk 4]
```diff
@@ -209,6 +212,17 @@ def _configure_cluster_management_service(message_bus_obj):
...
+        rgw_plugin_obj = rgw_plugin.RGWPlugin(
+            auth_user, auth_user_access_key, auth_user_secrete_key)
+        CsmRestApi._app[const.RGW_S3_USERS_SERVICE] = RgwUsersService(rgw_plugin_obj)
```
Comment (pawankumarsrivastava) [inline csm/core/agent/csm_agent.py:224]: Rename RgwUsersService --> S3IAMUserService

File: csm/core/controllers/__init__.py [hunk 5]
```diff
@@ -23,6 +23,7 @@
 from .cluster_management import ClusterOperationsView
 from .audit_log import AuditLogShowView, AuditLogDownloadView
 from .maintenance import MaintenanceView
+from .rgw.s3.users import RgwUserListView
```
Comment (pawankumarsrivastava) [inline csm/core/controllers/__init__.py:26]: Rename RgwUserListView -> S3IAMUserListView

File: csm/core/controllers/rgw/s3/users.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+import json
+from marshmallow import Schema, fields, ValidationError
+from cortx.utils.log import Log
+from csm.common.errors import InvalidRequest
+from csm.common.permission_names import Resource, Action
+from csm.core.blogic import const
+from csm.core.controllers.view import CsmView, CsmAuth
+from csm.core.controllers.validators import ValidationErrorFormatter
+
+class RgwUserCreateSchema(Schema):
```
Comment (pawankumarsrivastava) [inline csm/core/controllers/rgw/s3/users.py:25]: RgwUserCreateSchema -> UserCreateSchema

**Review by pawankumarsrivastava — CHANGES_REQUESTED** [review]
Remove RGW from class names wherever possible.

File: schema/roles.json [hunk 7]
```diff
@@ -32,7 +33,8 @@
             "replace_node": ["list"],
             "health": ["list"],
             "security": ["read"],
-			"cluster_management": ["create", "list"]
+            "cluster_management": ["create", "list"],
+            "s3_iam_users": ["create"]
```
Comment (avrnv) [inline schema/roles.json:37]: Question: we already have s3iamusers access rights: do we need new one?

**Review by avrnv — CHANGES_REQUESTED** [review]

File: csm/plugins/cortx/rgw.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    def _create_error(self, status: HTTPStatus, body: str) -> Any:
+        """
+        Converts a body of a failed query into IamError object.
+
+        :param status: HTTP Status code.
+        :param body: parsed HTTP response (dict) with the error's decription.
+        :returns: instance of error.
+        """
+
+        pass
```
[the remark this answers is not in the document]
Comment (rohitkolapkar) [reply csm/plugins/cortx/rgw.py]: fixed

File: csm/core/blogic/const.py [hunk 9]
```diff
@@ -923,3 +925,6 @@
 DB_CONSUL_CONFIG_PORT = 'databases>consul_db>config>port'
 DB_CONSUL_CONFIG_PASSWORD = 'databases>consul_db>config>password'
 DB_CONSUL_CONFIG_LOGIN = 'databases>consul_db>config>login'
+
+# RGW S3 API 
+CREATE_USER_OPERATION = 'CREATE_USER'
```
[the remark this answers is not in the document]
Comment (rohitkolapkar) [reply csm/core/blogic/const.py]: Fixed

File: csm/core/controllers/rgw/s3/users.py (hunk 6 above)
Comment (rohitkolapkar) [reply csm/core/controllers/rgw/s3/users.py]: Fixed

File: csm/core/controllers/__init__.py (hunk 5 above)
Comment (rohitkolapkar) [reply csm/core/controllers/__init__.py]: Fixed

File: csm/core/agent/csm_agent.py (hunk 4 above)
Comment (rohitkolapkar) [reply csm/core/agent/csm_agent.py]: Fixed

File: csm/core/agent/csm_agent.py (hunk 3 above)
Comment (rohitkolapkar) [reply csm/core/agent/csm_agent.py]: Fixed

File: csm/core/agent/csm_agent.py (hunk 2 above)
Comment (rohitkolapkar) [reply csm/core/agent/csm_agent.py]: Fixed

File: csm/common/permission_names.py (hunk 1 above)
Comment (rohitkolapkar) [reply csm/common/permission_names.py]: Fixed

File: schema/roles.json (hunk 7 above)
Comment (rohitkolapkar) [reply schema/roles.json:37]: Old- s3iamusers access right is related to s3 user role, here I'm giving permissions to admin and manage user role. 
Moreover In the upcoming sprints we are going to clean up "legacy-s3" code, so old one will get deleted.

Comment (rohitkolapkar) [reply schema/roles.json:37]: @avrnv can you please resolve this ? If you want, we can discuss on it.
Thanks

File: csm/core/controllers/rgw/s3/users.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    generate_key = fields.Bool(data_key=const.GENERATE_KEY, missing=None)
+    max_buckets = fields.Int(data_key=const.MAX_BUCKETS, missing=None)
+    suspended = fields.Bool(data_key=const.SUSPENDED, missing=None)
+    tenant = fields.Str(data_key=const.TENANT, missing=None)
+
+    @validates_schema
+    def invalidate_empty_values(self, data, **kwargs):
+        """This method invalidates the empty strings."""
+        for key, value in data.items():
+            if value is not None and not str(value).strip():
```
Comment (avrnv) [inline csm/core/controllers/rgw/s3/users.py:46]: Minor: could `validate=validate.Length(min=1)` do this job for you?

File: csm/core/data/models/rgw.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    KeyExists = "Provided access key exists and belongs to another user."
+    EmailExists = "Provided email address exists."
+    InvalidCapability = "Attempt to grant invalid admin capability."
+
+class RgwError:
+    """Class that describes a non-successful result"""
+
+    http_status: int
+    error_code: RgwErrors
+    error_message: str
```
Comment (avrnv) [inline csm/core/data/models/rgw.py:49]: Minor: add a blank line to the end of the file.

File: csm/plugins/cortx/rgw.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+    def __init__(self):
+        """
+        Initialize RGW plugin
+        """
+        config = CsmRgwConfigurationFactory.get_rgw_connection_config()
+        self._rgw_admin_client = RGWAdminClient(config.auth_user_access_key,
+            config.auth_user_secret_key, config.host, config.port)
+        Log.info(f"RGW admin uid: {const.ADMIN_UID}")
+        self._api_operations = {
```
Comment (avrnv) [inline csm/plugins/cortx/rgw.py:37]: Minor: could it be placed into a template file?

File: csm/core/controllers/rgw/s3/base.py [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        try:
+            yield None
+        except S3ServiceError as error:
+            raise CsmHttpException(error.status,
+                                   S3_SERVICE_ERROR,
+                                   error.code,
+                                   error.message,
+                                   error.message_args)
+        else:
+            return
```
[the remark this answers is not in the document]
Comment (rohitkolapkar) [reply csm/core/controllers/rgw/s3/base.py:40]: done

File: csm/core/data/models/rgw.py (hunk 11 above)
Comment (rohitkolapkar) [reply csm/core/data/models/rgw.py:49]: done

File: csm/plugins/cortx/rgw.py (hunk 12 above)
Comment (rohitkolapkar) [reply csm/plugins/cortx/rgw.py:37]: @shrimetta could you please look into this ?

File: csm/core/blogic/const.py [hunk 14]
```diff
@@ -923,3 +925,21 @@
...
 DB_CONSUL_CONFIG_PASSWORD = 'databases>consul_db>config>password'
 DB_CONSUL_CONFIG_LOGIN = 'databases>consul_db>config>login'
+
+# RGW AdminOps
+CREATE_USER_OPERATION = 'CREATE_USER'
+ADMIN_UID = 'admin'
+RGW_CLIENT_ERROR_MSG = 'Can not process the request'
+
+# RGW AdminOps- create-user
+UID = 'uid'
```
[the remark this answers is not in the document]
Comment (rohitkolapkar) [reply csm/core/blogic/const.py:935]: done

File: csm/core/controllers/rgw/s3/users.py (hunk 10 above)
Comment (rohitkolapkar) [reply csm/core/controllers/rgw/s3/users.py:46]: Actually No, as validate=validate.Length(min=1) allows {"uid":"   "}

File: csm/plugins/cortx/rgw.py (hunk 12 above)
Comment (rohitkolapkar) [reply csm/plugins/cortx/rgw.py]: @avrnv completed

**Review by soniyamoholkar — APPROVED** [review]

**Review by pawankumarsrivastava — APPROVED** [review]

**Review by shrimetta — APPROVED** [review]

**Review by avrnv — APPROVED** [review]

**Pull request merged**